### ISSUE TITLE
Signature command - multiple problems

### ISSUE THREAD

**Issue by calpaugh** [jay0lee/GAM#328]
I have encountered multiple problems with the signature command. It really does not appear to function at all like I expected. I was hoping to be able to use it to deploy custom signatures for all of my users. Here is a brief description of 3 problems.

problem 1:

Steps:

1. create a simple signature with 1, 2, 3 on 3 separate lines in google settings.

2. using the GAM command "show signature", the following html is returned:

<div dir="ltr">1<div>2</div><div>3</div><div><br></div></div>

3. If I now save this into  a file (sig.txt) and use GAM to set this as the signature for a user and then return to google settings for the user, i see the signature as follows:

ÿþ<div dir="ltr">1<div>2</div><div>3</div><div><br></div></div>
 
notice the extraneous characters and it wasn't interpreted as html.

Since this simple command fails, i felt like giving up. But here are two more problems:

problem 2:

When I used my own html, for example this snippet (saved as "sig.txt")

<table><tr><td>Title:</td><td>#Title#</td></tr></table>

And use the signature command to set the signature from "sig.txt" the resulting signature when viewed in google preferences has extra lines before and after the table.

Subsequently, using the GAM show signature command to retrieve the signature using GAM, it returns this html:

<table><td><br><tr><td>Title:</td><td>#Title#</td><td><br></td></tr><td><br></td></td></table><br><br>

Notice all of the extra <br> tokens that have been inserted. (the ones that are visible in google settings too)

problem 3:

The "replace" command does not work, even on the simplest example. It does not replace the text.

Using the sig.txt file from problem 2, i tried the following command:

PS C:\gam> gam user somebody@anywhere.com signature file sig.txt replace "#Title#" "Director"

This does not replace the text as expected. I am quite certain this is the notation given by Ross back in a June28 2016 posting.

So basically, the "signature" command does not work anywhere close to how I expected it to. Does anybody have success with this command? This looked so promising.

I am using GAM 4.0.3 on windows.

 Christopher

**Comment by taers232c** [comment]
Christopher,

Problem 3:
gam <UserTypeEntity> signature <String>|(file <FileName> [charset <Charset>]) (replace <Tag> <String>)*
Every instance of {Tag} in the signature/message will be replaced by <String>. Instances of the form {RT}...{Tag}...{/RT} will be eliminated
if there was no <Tag> specified that matches Tag or if a <Tag> matching Tag was specified but the matching <String> is empty.

Your sig.txt should be:
Title: {#Title#}

Problems 1 and 2: Please send me (ross.scroggs@gmail.com) your sig.txt file.

Ross
--
ross.scroggs@gmail.com

**Comment by taers232c** [comment]
Christoper,

Your sig.txt is encoded as UTF-16, the funny characters are the Unicode Byte Order Mark (BOM).
Say: gam user <user> signature file sig.txt charset utf-16
Or save sig.txt as a Text Document as opposed to a Unicode Text Document.

Ross
--
ross.scroggs@gmail.com

**Comment by calpaugh** [comment]
Ross sorted this out completely. Thanks Ross! Here are the notes that I made:

1. make sure the html source file encoding matches the encoding specified with the GAM command line

2. make sure that the template elements are formatted {#Name#}  and the command line ones "#Name" . You need the {} around each element to be substituted inside the html template

3. make sure your html is a single input line with no \n or \r characters; remember notepad.exe ignores \n alone; if a text file contains \n characters, notepad.exe will show it as a single line, but doing a CMD "cat" on the file will show that it has line breaks;  if you leave any \n characters in the html input file, then will be transformed to extra <br> elements by GAM.

With that, its all good! Nice utility! You can easily use this to deploy custom template-based signatures to an entire domain with two command lines. One powershell cmd to create a csv input file and one GAM command to push out the signatures in G-Suite. Now why is this not be a feature in G-Suite admin console? Its not like the information isn't available within the G-Suite space. Microsoft is no better; Exchange Online also lacks template-based corporate signatures.
